File: algorithms/cpp/最小覆盖子串/最小覆盖子串.cpp
// Source : https://leetcode.cn/problems/minimum-window-substring/description/
// Author : Yanan Wang
// Date   : 2023-08-06

/**
 * 本题先用暴力解法，然后优化细节就能变成动态规划的做法，但是在判断两字符串时，一般的做法是遍历t和s中的字符数量进行比较。
 * 上面的方法比较慢（官方题解），下面是更优解法：
 *    用cnt来记录窗口中“有效字符种类数量”，当cnt==t中的字符种类数量时，进行更新。
 *    1、 在进窗口后，判断该字符在窗口和t中的数量，若相等，则为有效字符，++cnt
 *    2、 在出窗口前，判断该字符在窗口和t中的数量，若相等，则该有效字符被剔除掉了，--cnt
 */

class Solution {
public:
    string minWindow(string s, string t) {
        char base[128] = { 0 }, kinds = 0;
        for (const auto &c : t) 
            if (0 == base[c]++) ++kinds;
        int start = 0, len = INT_MAX;
        char window[128] = { 0 };
        for (int left = 0, right = 0, cnt = 0; right < s.size(); ++right) {
            char in = s[right];
            if (base[in] == ++window[in]) ++cnt;
            while (cnt == kinds) {
                if (right - left + 1 < len) {
                    len = right - left + 1;
                    start = left;
                }
                char out = s[left++];
                if (base[out] == window[out]--) --cnt;
            }
        }
        return len == INT_MAX ? "" : s.substr(start, len);
    }
};

File: algorithms/python/最小覆盖子串/最小覆盖子串.py
# Source : https://leetcode.cn/problems/M1oyTv/description/
# Author : Yanan Wang
# Date   : 2023-12-25

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        kind = 0
        base = [0] * 128
        for elem in t:
            if base[ord(elem)] == 0:
                kind += 1
            base[ord(elem)] += 1

        ans_begin = 0
        ans_end = -1
        ans_len = float('inf')
        left = right = cnt = 0
        win = [0] * 128
        while right < len(s):
            win[ord(s[right])] += 1            
            if win[ord(s[right])] == base[ord(s[right])]:
                cnt += 1
            while cnt == kind:
                if ans_len > right - left + 1:
                    ans_len = right - left + 1
                    ans_begin = left
                    ans_end = right
                if win[ord(s[left])] == base[ord(s[left])]:
                    cnt -= 1
                win[ord(s[left])] -= 1
                left += 1
            right += 1    
            
        return s[ans_begin: ans_end + 1] if ans_len != float('inf') else ""     
